# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: Mai Quý Trung
- MSSV: 20127370
- Lớp: 20CLC11

# Import

In [14]:
import pandas as pd
import numpy as np
from sklearn import model_selection

# Đọc dữ liệu

In [15]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
X_train = train.iloc[:, :-1]    # Dataframe (chứa 10 đặc trưng huấn luyện)
y_train = train.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

X_test = test.iloc[:, :-1]      # Dataframe (chứa 10 đặc trưng kiểm tra)
y_test = test.iloc[:, -1]       # Series    (chứa 1 giá trị mục tiêu kiểm tra)

# Sinh viên có thể sử dụng các khác nếu cần

In [16]:
print("Training file:")
train.head()

Training file:


Adult Mortality   BMI  Polio  Diphtheria  HIV/AIDS         GDP  \
0            268.0  18.1   62.0        64.0       0.1  631.744976   
1            272.0  17.6   67.0        67.0       0.1  669.959000   
2            275.0  17.2   68.0        68.0       0.1   63.537231   
3            279.0  16.7   66.0        66.0       0.1  553.328940   
4            281.0  16.2   63.0        63.0       0.1  445.893298   

   Thinness age 10-19  Thinness age 5-9  Income composition of resources  \
0                17.7              17.7                            0.470   
1                17.9              18.0                            0.463   
2                18.2              18.2                            0.454   
3                18.4              18.4                            0.448   
4                18.6              18.7                            0.434   

   Schooling  Life expectancy  
0        9.9             59.9  
1        9.8             59.5  
2        9.5             59.2  
3        9.2             58.8  
4        8.9             58.6

In [17]:
print("Testing file:")
test.head()

Testing file:


Adult Mortality   BMI  Polio  Diphtheria  HIV/AIDS          GDP  \
0            263.0  19.1    6.0        65.0       0.1   584.259210   
1            271.0  18.6   58.0        62.0       0.1   612.696514   
2             74.0  58.0   99.0        99.0       0.1  3954.227830   
3              8.0  57.2   98.0        98.0       0.1  4575.763787   
4             11.0  58.4   95.0        95.0       0.1   547.851700   

   Thinness age 10-19  Thinness age 5-9  Income composition of resources  \
0                17.2              17.3                            0.479   
1                17.5              17.5                            0.476   
2                 1.2               1.3                            0.762   
3                 1.2               1.3                            0.761   
4                 6.0               5.8                            0.741   

   Schooling  Life expectancy  
0       10.1             65.0  
1       10.0             59.9  
2       14.2             77.8  
3       14.2             77.5  
4       14.4             75.4

# Cài đặt hàm

In [18]:
# Cài đặt các hàm cần thiết ở đây
class OLSLinearRegression:
    def fit(self, X, y):
        X_pinv = np.linalg.inv(X.T @ X) @ X.T
        self.w = X_pinv @ y
        return self

    def get_params(self):
        return self.w

    def predict(self, X):
        return np.sum(self.w.ravel() * X, axis=1)  

def model_mse(y_test, y_pred):
    return np.mean((y_test.ravel() - y_pred.ravel())**2)

def model_rmse(y_test, y_pred):
    return np.sqrt(model_mse(y_test, y_pred))


# Yêu cầu 1a: Sử dụng toàn bộ 10 đặc trưng đề bài cung cấp (2 điểm) 

In [19]:
# Phần code cho yêu cầu 1a
def train_all_feature(X_train, y_train, X_test):
    w = OLSLinearRegression().fit(X_train, y_train)
    y_pred = w.predict(X_test)
    return y_pred, w

In [20]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
def task_1a(X_train, y_train, X_test, y_test):
    y_pred, w = train_all_feature(X_train, y_train, X_test)
    return model_rmse(y_test, y_pred), np.array(w.get_params()).reshape(-1, 1)

rmse, w = task_1a(X_train.to_numpy(), y_train.to_numpy(), X_test.to_numpy(), y_test.to_numpy())
print("Weight:\n", w, "\nRMSE:", rmse)

Weight:
 [[ 1.51013627e-02]
 [ 9.02199807e-02]
 [ 4.29218175e-02]
 [ 1.39289117e-01]
 [-5.67332827e-01]
 [-1.00765115e-04]
 [ 7.40713438e-01]
 [ 1.90935798e-01]
 [ 2.45059736e+01]
 [ 2.39351661e+00]] 
RMSE: 7.0640464305843516


Công thức hồi quy

$$\text{Life expectancy} = 0.0151013627 * \text{(Adult Mortality)} + 0.0902199807 * \text{(BMI)} \\+ 0.0429218175 * \text{(Polio)} + 0.139289117 * \text{(Diphtheria)} \\ - 0.567332827 * \text{(HIV/AIDS)} - 0.000100765115 * \text{(GDP)} \\+ 0.740713438 * \text{(Thinness age 10-19)} + 0.190935798 * \text{(Thinness age 5-9)} \\+ 24.5059736 * \text{(Income composition of resources)} + 2.39351661 * \text{(Schooling)}

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng, tìm mô hình cho kết quả tốt nhất (2 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [21]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu
category = ["Adult Mortality", "BMI", "Polio", "Diphtheria", "HIV/AIDS", "GDP", "Thinness age 10-19", "Thinness age 5-9", "Income composition of resources", "Schooling", "Life expectancy"]

def task_1b(train, fold = 5):

    w_arr = np.zeros(train.shape[1] - 1)
    rmse_arr = np.zeros(train.shape[1] - 1)
    shuffle_data =  model_selection.KFold(n_splits=fold, shuffle=True)

    for train_split, test_split in shuffle_data.split(train):

        idx = 0
        
        for column in category:

            if column != 'Life expectancy':
                
                feature_train = train.iloc[train_split].loc[:,[column]]
                label_train = train.iloc[train_split].loc[:,['Life expectancy']]

                feature_test = train.iloc[test_split].loc[:,[column]]
                label_test = train.iloc[test_split].loc[:,['Life expectancy']]
                    
                rmse, w = task_1a(feature_train.to_numpy(), label_train.to_numpy(), feature_test.to_numpy(), label_test.to_numpy())

                rmse_arr[idx] += rmse
                w_arr[idx] += w
                idx += 1

    w_arr /= fold
    rmse_arr /= fold

    print('RMSE:\n', rmse_arr.reshape(-1, 1))
    print('Weight:\n', w_arr.reshape(-1, 1))

    return rmse_arr, category[np.argmin(rmse_arr)]

In [22]:
rmse, best_1b = task_1b(train)
print('Min RMSE:', np.min(rmse), '\nBest feature:', best_1b)


RMSE:
 [[46.20269332]
 [27.91251384]
 [17.93853784]
 [16.06936343]
 [67.0583796 ]
 [60.20524097]
 [52.04589701]
 [51.8434712 ]
 [13.20208735]
 [11.7866964 ]]
Weight:
 [[2.47765822e-01]
 [1.50581054e+00]
 [7.85637083e-01]
 [7.84449459e-01]
 [3.58466887e+00]
 [3.02679343e-03]
 [6.94583391e+00]
 [6.88993932e+00]
 [1.04706096e+02]
 [5.55741166e+00]]
Min RMSE: 11.78669639939751 
Best feature: Schooling


In [23]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_feature_model
feature_train_1b = train.loc[:, [best_1b]]
feature_test_1b = test.loc[:, [best_1b]]

rmse_1b, w_1b = task_1a(feature_train_1b.to_numpy(), y_train.to_numpy(), feature_test_1b.to_numpy(), y_test.to_numpy())
print("New RMSE:", rmse_1b, '\nNew Weight:', w_1b)

New RMSE: 10.26095039165537 
New Weight: [[5.5573994]]


Công thức hồi quy

$$\text{Life expectancy} = 5.5573994 * \text{(Schooling)}

# Yêu cầu 1c: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

In [24]:
# Phần code cho yêu cầu 1c
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu
# ['BMI', 'Polio', "Diphtheria", "HIV/AIDS", 'GDP', 'Schooling', 'Life expectancy']
m = 7

def build_data(train_data, opt):
    if opt == 0:
        return train_data.loc[:, ["Thinness age 10-19", "Polio", "Diphtheria", "Thinness age 5-9", "Income composition of resources", "Schooling", "Life expectancy"]]
    if opt == 1:
        return train_data.loc[:, ["BMI", "GDP", "HIV/AIDS", "Schooling", "Life expectancy"]]
    if opt == 2:
        return train_data.loc[:, ["BMI", "GDP", "Polio", "Life expectancy"]]
    if opt == 3:
        return pd.concat([train_data.loc[:, ['Schooling']], pd.DataFrame([[data[i] ** 2 if i != 1 else data[i] for i in range(2)] for data in ((train_data.loc[:, ['Schooling', 'Life expectancy']]).to_numpy())], columns = ['Schooling', 'Life expectancy'])], axis=1, join='inner')
    if opt == 4:
        return pd.DataFrame([[data[i] ** 2 if i != 10 else data[i] for i in range(train_data.shape[1])] for data in train_data.to_numpy()], columns = category)
    if opt == 5:
        return pd.DataFrame([[data[i] ** 3 if i != 10 else data[i] for i in range(train_data.shape[1])] for data in train_data.to_numpy()], columns = category)
    if opt == 6:
        return pd.DataFrame([[data[i] ** 2 if i != 1 else data[i] for i in range(2)] for data in ((train_data.loc[:, ['Schooling', 'Life expectancy']]).to_numpy())], columns = ['Schooling', 'Life expectancy'])
    else:
        return train_data

def task_1c(train, fold = 5):

    w_arr = np.zeros((m, 10))
    rmse_arr = np.zeros((m, 1))
    shuffle_data =  model_selection.KFold(n_splits=fold, shuffle=True)

    for train_split, test_split in shuffle_data.split(train):
    
        for idx in range(m):

            model = build_data(train, idx)

            feature_train = model.iloc[train_split].iloc[:, :-1]
            label_train = model.iloc[train_split].loc[:,['Life expectancy']]

            feature_test = model.iloc[test_split].iloc[:, :-1]
            label_test = model.iloc[test_split].loc[:,['Life expectancy']]
                            
            rmse, w = task_1a(feature_train.to_numpy(), label_train.to_numpy(), feature_test.to_numpy(), label_test.to_numpy())

            for _ in range(10 - len(w)):
                w = np.append(w, float(0))

            rmse_arr[idx] += rmse
            w_arr[idx] += w.reshape(-1)
            
    w_arr /= fold
    rmse_arr /= fold

    print('RMSE:\n', rmse_arr)
    print('Weight:\n', w_arr)
    return rmse_arr, np.argmin(rmse_arr)

rmse_arr, best_1c = task_1c(train)
print("My best model:", best_1c + 1)
print(build_data(train, best_1c))


RMSE:
 [[ 8.32959999]
 [11.15332909]
 [15.59313433]
 [ 6.90659883]
 [13.68429193]
 [19.12347377]
 [24.49011507]]
Weight:
 [[ 7.86991922e-01  5.40719301e-02  1.57832699e-01  1.30347306e-01
   2.93232326e+01  2.30170242e+00  0.00000000e+00  0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 1.10603309e-02 -4.13243269e-04 -2.15027181e-01  5.76191531e+00
   0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 4.44798658e-01  1.21418769e-04  5.83154354e-01  0.00000000e+00
   0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 9.25996429e+00 -2.76796991e-01  0.00000000e+00  0.00000000e+00
   0.00000000e+00  0.00000000e+00  0.00000000e+00  0.00000000e+00
   0.00000000e+00  0.00000000e+00]
 [ 1.04797760e-04  1.49018306e-03  9.50139326e-04  2.89801778e-03
  -2.57367892e-02 -5.10932568e-09  5.40335332e-02  1.09159211e-02
   5.35380307e+01  3.88399412e-02]
 [ 2.27912740e-07  2.49567860e-05  1.285630

In [25]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện
feature_train_1c = build_data(train, best_1c).iloc[:, :-1]
feature_test_1c = build_data(test, best_1c).iloc[:, :-1]

rmse_1c, w_1c = task_1a(feature_train_1c.to_numpy(), y_train.to_numpy(), feature_test_1c.to_numpy(), y_test.to_numpy())

In [26]:
# Gọi hàm RMSE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model
print('New Weight:', w_1c, '\nNew RMSE:', rmse_1c)

New Weight: [[ 9.25853343]
 [-0.27667035]] 
New RMSE: 5.9986721709342445


Công thức hồi quy

$$\text{Life expectancy} = 9.25853343 * \text{(Schooling)} - 0.27667035 * \text{(Schooling} ^ 2)